qemu/cpu/code_generation.py

```diff
@@ -0,0 +1,1329 @@
+                        ",@s".join(["%lx"] * len(operands))
+                    ),
+                    *operands
+                )
+            )
+
+        if instruction.branch and not DEBUG_DECODER:
+            node(
+                OpAssign(
+                    OpSDeref(ctx, "bstate"),
+                    Type["BS_BRANCH"]
+                )
+            )
+
+        node(OpAssign(result, total_read))
+
+    default_switch_case_nodes = [
+        Call("set_pc", ctx_pc),
+        Call("gen_helper_illegal", cpu_env),
+        OpAssign(
+            OpSDeref(ctx, "bstate"),
+            Type["BS_EXCP"]
+        )
+    ]
+
+    ParseTreeCodeBuilder(cputype, body, gen_field_read, decode_opc_epilogue,
+        default_switch_case_nodes
+    )
+
+    body(Return(result))
+
+def fill_disas_set_info_body(cputype, function):
+    function.body = BodyTree()(
+        OpAssign(
+            OpSDeref(function.args[1], "mach"),
+            getattr(Type["bfd_architecture"], cputype.bfd_arch_name)
+        ),
+        OpAssign(
+            OpSDeref(function.args[1], "print_insn"),
+            Type[cputype.print_insn_name]
+        )
+    )
+
+def fill_disas_write_helper_body(function):
+    function.body = BodyTree()(Return(0))
+
+def fill_dump_state_body(cputype, function, reg_vars):
+    function.body = body = BodyTree()
+
+    cpu = Pointer(Type[cputype.struct_instance_name])("cpu")
+    env = Pointer(Type[cputype.struct_name])("env")
+
+    out_file = function.args[1]
+    if get_vp("dump_state has cpu_fprintf argument"):
+        fprintf_func = function.args[2]
+    else:
+        fprintf_func = Type["qemu_fprintf"]
+    i = Type["int"]("i")
+
+    body(
+        Declare(
+            OpDeclareAssign(
+                cpu,
+                MCall(cputype.qtn.for_macros, function.args[0])
+            )
+        ),
+        Declare(
+            OpDeclareAssign(
+                env,
+                OpAddr(OpSDeref(cpu, "env"))
+            )
+        ),
+        Declare(i),
+        NewLine()
+    )
+
+    for r, _, names_array in reg_vars:
+        if r.bank_size:
+            body(
+                LoopFor(OpAssign(i, 0), OpLess(i, r.bank_size), OpPreInc(i))(
+                    Call(
+                        fprintf_func,
+                        out_file,
+                        "%s=0x%08x",
+                        OpIndex(names_array, i),
+                        OpIndex(OpSDeref(env, r.name), i)
+                    ),
+                    BranchIf(OpEq(OpRem(i, 4), 3))(
+                        Call(fprintf_func, out_file, "\\n"),
+                        BranchElse()(Call(fprintf_func, out_file, ' '))
+                    )
+                )
+            )
+        else:
+            body(
+                Call(
+                    fprintf_func,
+                    out_file,
+                    r.name + "=0x%08x\\n",
+                    OpSDeref(env, r.name)
+                )
+            )

```

Comment:
That does not prints `\n` if number of registers is not multiple of 4.

```
pc=0x0000640e
r1=0x00005bfe r2=0x00000000 r3=0x00000000 r4=0x00000000
r5=0x00000000 r6=0x00000000 r7=0x00000000 r8=0x00000000
r9=0x00000000 r10=0x00000000 r11=0x00000000 r12=0x00000000
r13=0x00000000 r14=0x00000000 r15=0x00000000 IN: main
0x0000647c:  MOV.W	 , r9
0x0000647e:  SUB.W	 , r9
```

(`r15=0x00000000 IN: main`)

```
    cpu_fprintf(f, "pc=0x%08x\n", env->pc);
    for (i = 0; i < 15; ++i) {
        cpu_fprintf(f, "%s=0x%08x", regs_names[i], env->regs[i]);
        if (i % 4 == 3) {
            cpu_fprintf(f, "\n");
        } else {
            cpu_fprintf(f, " ");
        }
    }
```

Reply:
This also adds unnecessary space after last reg in that case.

Reply:
It's a fix

```
if r.bank_size % 4:
    body(Call(fprintf_func, out_file, "\\n"))
```